File: services/guide.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class TranscriptSegment:
    start: float
    end: float
    text: str
    confidence: float = 0.85


@dataclass(frozen=True)
class VisualEvent:
    start: float
    end: float
    description: str
    screenshot: str | None = None
    confidence: float = 0.75


def synthesize_walkthrough(
    transcript: dict[str, Any],
    visual: dict[str, Any],
    options: dict[str, Any] | None = None,
) -> dict[str, Any]:
    options = options or {}
    max_steps = int(options.get("maxSteps", 12))
    segments = [_segment(item) for item in transcript.get("segments", [])]
    events = [_event(item) for item in visual.get("events", [])]

    if not segments and not events:
        return {
            "title": "Untitled Walkthrough",
            "summary": "No transcript or visual events were detected.",
            "steps": [],
            "warnings": ["The analysis did not produce usable transcript or visual events."],
        }

    timeline = events or [
        VisualEvent(
            start=segment.start,
            end=segment.end,
            description=segment.text,
            confidence=segment.confidence,
        )
        for segment in segments
    ]

    steps: list[dict[str, Any]] = []
    warnings: list[str] = []

    for event in timeline[:max_steps]:
        aligned = _overlapping_segments(event, segments)
        narration = " ".join(segment.text.strip() for segment in aligned if segment.text.strip())
        heading = _heading_from(narration or event.description)
        instruction = _instruction_from(narration, event.description)
        confidence_values = [event.confidence, *[segment.confidence for segment in aligned]]
        confidence = sum(confidence_values) / len(confidence_values)
        if confidence < 0.55:
            warnings.append(f"Review step {len(steps) + 1}; confidence is low.")

        steps.append(
            {
                "index": len(steps) + 1,
                "heading": heading,
                "instruction": instruction,
                "timestampStart": round(event.start, 2),
                "timestampEnd": round(event.end, 2),
                "screenshot": event.screenshot,
                "evidence": _evidence(narration, event.description),
                "confidence": round(max(0.0, min(confidence, 1.0)), 2),
            }
        )

    title = _heading_from(segments[0].text if segments else events[0].description)
    summary = _summary(segments, events)

    if not segments:
        warnings.append("No narration transcript was available; steps are based on visual analysis only.")
    if not events:
        warnings.append("No visual events were available; steps are based on narration only.")

    return {"title": title, "summary": summary, "steps": steps, "warnings": warnings}


def render_markdown(walkthrough: dict[str, Any]) -> str:
    lines = [f"# {walkthrough['title']}", "", walkthrough.get("summary", ""), ""]
    for step in walkthrough.get("steps", []):
        lines.extend([f"## {step['index']}. {step['heading']}", ""])
        if step.get("screenshot"):
            lines.extend([f"![{step['heading']}](assets/{step['screenshot']})", ""])
        lines.extend(
            [
                step["instruction"],
                "",
                f"Timestamp: {_timestamp(step['timestampStart'])} - {_timestamp(step['timestampEnd'])}",
                f"Confidence: {round(step['confidence'] * 100)}%",
            ]
        )
        if step.get("evidence"):
            lines.append(f"Evidence: {step['evidence']}")
        lines.append("")

    warnings = walkthrough.get("warnings", [])
    if warnings:
        lines.extend(["## Review Notes", ""])
        lines.extend(f"- {warning}" for warning in warnings)
        lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def _segment(item: dict[str, Any]) -> TranscriptSegment:
    return TranscriptSegment(
        start=float(item.get("start", 0)),
        end=float(item.get("end", item.get("start", 0))),
        text=str(item.get("text", "")).strip(),
        confidence=float(item.get("confidence", 0.85)),
    )


def _event(item: dict[str, Any]) -> VisualEvent:
    return VisualEvent(
        start=float(item.get("start", 0)),
        end=float(item.get("end", item.get("start", 0))),
        description=str(item.get("description", "")).strip(),
        screenshot=item.get("screenshot"),
        confidence=float(item.get("confidence", 0.75)),
    )


def _overlapping_segments(event: VisualEvent, segments: list[TranscriptSegment]) -> list[TranscriptSegment]:
    return [
        segment
        for segment in segments
        if segment.start <= event.end and segment.end >= event.start
    ]


def _heading_from(text: str) -> str:
    words = [word.strip(".,:;!?") for word in text.split() if word.strip(".,:;!?")]
    if not words:
        return "Review this step"
    compact = " ".join(words[:7])
    return compact[:1].upper() + compact[1:]


def _instruction_from(narration: str, description: str) -> str:
    if narration and description:
        return f"{narration.strip()} The screen shows: {description.strip()}"
    if narration:
        return narration.strip()
    return f"Follow the visible action: {description.strip()}"


def _evidence(narration: str, description: str) -> str:
    parts = []
    if narration:
        parts.append(f"narration: {narration}")
    if description:
        parts.append(f"visual: {description}")
    return "; ".join(parts)


def _summary(segments: list[TranscriptSegment], events: list[VisualEvent]) -> str:
    if segments:
        text = " ".join(segment.text for segment in segments[:3]).strip()
        return text if text else "Generated from screen recording narration and visual events."
    return events[0].description if events else "Generated from screen recording analysis."


def _timestamp(seconds: float) -> str:
    safe = max(0, int(seconds))
    return f"{safe // 60:02d}:{safe % 60:02d}"

File: src/lib/guide.ts
export type WalkthroughStep = {
  index: number;
  heading: string;
  instruction: string;
  timestampStart: number;
  timestampEnd: number;
  screenshot?: string;
  evidence: string;
  confidence: number;
};

export type Walkthrough = {
  title: string;
  summary: string;
  steps: WalkthroughStep[];
  warnings: string[];
};

export type TranscriptResponse = {
  segments: Array<{ start: number; end: number; text: string; confidence: number }>;
  duration?: number;
};

export type VisualResponse = {
  events: Array<{
    start: number;
    end: number;
    description: string;
    screenshot?: string;
    screenshotData?: string;
    confidence: number;
  }>;
};

export function synthesizeWalkthrough(
  transcript: TranscriptResponse,
  visual: VisualResponse,
  options: { maxSteps?: number } = {}
): Walkthrough {
  const maxSteps = options.maxSteps ?? 12;
  const segments = transcript.segments ?? [];
  const events = visual.events ?? [];

  if (segments.length === 0 && events.length === 0) {
    return {
      title: "Untitled Walkthrough",
      summary: "No transcript or visual events were detected.",
      steps: [],
      warnings: ["The analysis did not produce usable transcript or visual events."]
    };
  }

  const timeline: VisualResponse["events"] =
    events.length > 0
      ? events
      : segments.map((segment) => ({
          start: segment.start,
          end: segment.end,
          description: segment.text,
          confidence: segment.confidence
        }));

  const warnings: string[] = [];
  const steps: WalkthroughStep[] = timeline.slice(0, maxSteps).map((event, index) => {
    const aligned = segments.filter((segment) => segment.start <= event.end && segment.end >= event.start);
    const narration = aligned.map((segment) => segment.text.trim()).filter(Boolean).join(" ");
    const confidenceValues = [event.confidence, ...aligned.map((segment) => segment.confidence)];
    const confidence = confidenceValues.reduce((sum, value) => sum + value, 0) / confidenceValues.length;
    if (confidence < 0.55) warnings.push(`Review step ${index + 1}; confidence is low.`);

    return {
      index: index + 1,
      heading: headingFrom(narration || event.description),
      instruction: instructionFrom(narration, event.description),
      timestampStart: roundTime(event.start),
      timestampEnd: roundTime(event.end),
      screenshot: event.screenshot,
      evidence: evidenceFrom(narration, event.description),
      confidence: Math.max(0, Math.min(roundTime(confidence), 1))
    };
  });

  if (segments.length === 0) warnings.push("No narration transcript was available; steps are based on visual analysis only.");
  if (events.length === 0) warnings.push("No visual events were available; steps are based on narration only.");

  return {
    title: headingFrom(segments[0]?.text || events[0]?.description || "Untitled Walkthrough"),
    summary: summaryFrom(segments, events),
    steps,
    warnings
  };
}

export function buildMarkdown(walkthrough: Walkthrough): string {
  const lines: string[] = [`# ${walkthrough.title}`, "", walkthrough.summary, ""];

  for (const step of walkthrough.steps) {
    lines.push(`## ${step.index}. ${step.heading}`);
    lines.push("");
    if (step.screenshot) {
      lines.push(`![${step.heading}](assets/${step.screenshot})`);
      lines.push("");
    }
    lines.push(step.instruction);
    lines.push("");
    lines.push(`Timestamp: ${formatTimestamp(step.timestampStart)} - ${formatTimestamp(step.timestampEnd)}`);
    lines.push(`Confidence: ${Math.round(step.confidence * 100)}%`);
    if (step.evidence) {
      lines.push(`Evidence: ${step.evidence}`);
    }
    lines.push("");
  }

  if (walkthrough.warnings.length > 0) {
    lines.push("## Review Notes");
    lines.push("");
    walkthrough.warnings.forEach((warning) => lines.push(`- ${warning}`));
    lines.push("");
  }

  return lines.join("\n").trimEnd() + "\n";
}

export function formatTimestamp(seconds: number): string {
  const safeSeconds = Math.max(0, Math.floor(seconds));
  const minutes = Math.floor(safeSeconds / 60).toString().padStart(2, "0");
  const rest = (safeSeconds % 60).toString().padStart(2, "0");
  return `${minutes}:${rest}`;
}

function headingFrom(text: string): string {
  const words = text
    .split(/\s+/)
    .map((word) => word.replace(/[.,:;!?]+$/g, ""))
    .filter(Boolean);
  if (words.length === 0) return "Review this step";
  const compact = words.slice(0, 7).join(" ");
  return compact.slice(0, 1).toUpperCase() + compact.slice(1);
}

function instructionFrom(narration: string, description: string): string {
  if (narration && description) return `${narration.trim()} The screen shows: ${description.trim()}`;
  if (narration) return narration.trim();
  return `Follow the visible action: ${description.trim()}`;
}

function evidenceFrom(narration: string, description: string): string {
  return [
    narration ? `narration: ${narration}` : "",
    description ? `visual: ${description}` : ""
  ]
    .filter(Boolean)
    .join("; ");
}

function summaryFrom(
  segments: TranscriptResponse["segments"],
  events: VisualResponse["events"]
): string {
  if (segments.length > 0) {
    return segments
      .slice(0, 3)
      .map((segment) => segment.text)
      .join(" ")
      .trim() || "Generated from screen recording narration and visual events.";
  }
  return events[0]?.description || "Generated from screen recording analysis.";
}

function roundTime(value: number): number {
  return Math.round(value * 100) / 100;
}
